JDK-8260485: Simplify and unify handler vectors in Posix signal code

tstuefe opened the pull request:
In signal handling code, we have code sections which save signal handler state into vectors of sigaction structures, or of integers (if only flags are saved). All these code sections can be unified, disentangled and the using code simplified.

There are three places where we do this:

1) When installing hotspot signal handlers, should we find a handler in place and signal chaining is enabled, we save the original handler inside a sigaction array and a corresponding sigset:
https://github.com/openjdk/jdk/blob/e0d748d56f817b4fb5158aae86aee12d90b36356/src/hotspot/os/posix/signals_posix.cpp#L85
https://github.com/openjdk/jdk/blob/e0d748d56f817b4fb5158aae86aee12d90b36356/src/hotspot/os/posix/signals_posix.cpp#L338

2) if diagnostics are enabled with -Xcheck:jni, we periodically check if our hotspot signal handlers had been replaced (`static void check_signal_handler(int sig)`):
https://github.com/openjdk/jdk/blob/e0d748d56f817b4fb5158aae86aee12d90b36356/src/hotspot/os/posix/signals_posix.cpp#L766
To do that, we store information about the handlers we installed and we expect to be intact; in this case we only store the sigaction flags (`int sigflags[NSIG];`) and deduce the handler address from context.

3) There is a complicated dance between VMError and the posix signal handler code: If a fatal error happens, we enter error reporting and install the secondary handler (`VMError::install_secondary_signal_handler()`). Before doing that, we store the handler we replace in yet another array, in this case one array for the handler address, one for the flag:
https://github.com/openjdk/jdk/blob/e0d748d56f817b4fb5158aae86aee12d90b36356/src/hotspot/os/posix/vmError_posix.cpp#L77
I believe the purpose of this is to - when printing signal handlers as part of error reporting - print the original signal handler instead of the secondary crash handler (see `PosixSignals::print_signal_handler()`):
https://github.com/openjdk/jdk/blob/e0d748d56f817b4fb5158aae86aee12d90b36356/src/hotspot/os/posix/signals_posix.cpp#L1372
and additionally to not trip this warning here:
https://github.com/openjdk/jdk/blob/e0d748d56f817b4fb5158aae86aee12d90b36356/src/hotspot/os/posix/signals_posix.cpp#L1391

------

Changes in this patch:

- I added some convenience macros to check if a handler matches a given function (HANDLER_IS), check if a handler is set to ignore or default or both (HANDLER_IS_IGN, HANDLER_IS_DFL, HANDLER_IS_IGN_OR_DFL). Makes code more readable.
- I added convenience class `SavedSignalHandlers` to keep a vector of handler information by signal number.
- I used that class to cover cases (1)..(3):
	- `chained_handlers` contains all information of chained handlers
	- `expected_handlers` contains a copy of the handlers the hotspot installed
	- `replaced_handlers` contains information about replaced handlers

- about (1): I store the chained signal handler information in `chained_handlers` when installing a hotspot handler, UseSignalChaining is 1, and a non-default handler was encountered.

- about (2): I simplified the signal checking mechanism quite a bit: it compares the handler (address and flags) it finds present with expectations. Before this patch, the expected handler address was deduced in a hard-wired way, now, we just compare the active sigaction structure with the one we installed on VM start.

- about (3): when installing any handler (hotspot as well as user defined via java), I store the handler it replaced in `replaced_handlers`. I use that to print which handler had been replaced in `PosixSignals::print_signal_handler`. I simplified `PosixSignals::print_signal_handler` such that it does not retain any knowledge about hotspot signal handlers. Now, it just prints out the currently established handlers. In addition to that, it prints out chaining information and which handlers had been replaced. I removed the associated coding from VMError.

Output Before:
```
663 Signal Handlers:
664    SIGSEGV: javaSignalHandler in libjvm.so, sa_mask[0]=11100100010111111101111111111110, sa_flags=SA_RESTART|SA_SIGINFO
665     SIGBUS: javaSignalHandler in libjvm.so, sa_mask[0]=11100100010111111101111111111110, sa_flags=SA_RESTART|SA_SIGINFO
666     SIGFPE: javaSignalHandler in libjvm.so, sa_mask[0]=11100100010111111101111111111110, sa_flags=SA_RESTART|SA_SIGINFO
667    SIGPIPE: javaSignalHandler in libjvm.so, sa_mask[0]=11100100010111111101111111111110, sa_flags=SA_RESTART|SA_SIGINFO
668    SIGXFSZ: javaSignalHandler in libjvm.so, sa_mask[0]=11100100010111111101111111111110, sa_flags=SA_RESTART|SA_SIGINFO
669     SIGILL: javaSignalHandler in libjvm.so, sa_mask[0]=11100100010111111101111111111110, sa_flags=SA_RESTART|SA_SIGINFO
670    SIGUSR2: SR_handler in libjvm.so, sa_mask[0]=00000000000000000000000000000000, sa_flags=SA_RESTART|SA_SIGINFO
671     SIGHUP: UserHandler in libjvm.so, sa_mask[0]=11100100010111111101111111111110, sa_flags=SA_RESTART|SA_SIGINFO
672     SIGINT: UserHandler in libjvm.so, sa_mask[0]=11100100010111111101111111111110, sa_flags=SA_RESTART|SA_SIGINFO
673    SIGTERM: UserHandler in libjvm.so, sa_mask[0]=11100100010111111101111111111110, sa_flags=SA_RESTART|SA_SIGINFO
674    SIGQUIT: UserHandler in libjvm.so, sa_mask[0]=11100100010111111101111111111110, sa_flags=SA_RESTART|SA_SIGINFO
675    SIGTRAP: javaSignalHandler in libjvm.so, sa_mask[0]=11100100010111111101111111111110, sa_flags=SA_RESTART|SA_SIGINFO
```

Now:
```
 Signal Handlers:
    SIGSEGV: crash_handler in libjvm.so, mask=11100100010111111101111111111110, flags=SA_RESTART|SA_SIGINFO
   replaced:    SIGSEGV: javaSignalHandler in libjvm.so, mask=11100100010111111101111111111110, flags=SA_RESTART|SA_SIGINFO
     SIGBUS: crash_handler in libjvm.so, mask=11100100010111111101111111111110, flags=SA_RESTART|SA_SIGINFO
   replaced:     SIGBUS: javaSignalHandler in libjvm.so, mask=11100100010111111101111111111110, flags=SA_RESTART|SA_SIGINFO
     SIGFPE: crash_handler in libjvm.so, mask=11100100010111111101111111111110, flags=SA_RESTART|SA_SIGINFO
   replaced:     SIGFPE: javaSignalHandler in libjvm.so, mask=11100100010111111101111111111110, flags=SA_RESTART|SA_SIGINFO
    SIGPIPE: javaSignalHandler in libjvm.so, mask=11100100010111111101111111111110, flags=SA_RESTART|SA_SIGINFO
    SIGXFSZ: javaSignalHandler in libjvm.so, mask=11100100010111111101111111111110, flags=SA_RESTART|SA_SIGINFO
     SIGILL: crash_handler in libjvm.so, mask=11100100010111111101111111111110, flags=SA_RESTART|SA_SIGINFO
   replaced:     SIGILL: javaSignalHandler in libjvm.so, mask=11100100010111111101111111111110, flags=SA_RESTART|SA_SIGINFO
    SIGUSR2: SR_handler in libjvm.so, mask=00000000000000000000000000000000, flags=SA_RESTART|SA_SIGINFO
     SIGHUP: UserHandler in libjvm.so, mask=11100100010111111101111111111110, flags=SA_RESTART|SA_SIGINFO
     SIGINT: UserHandler in libjvm.so, mask=11100100010111111101111111111110, flags=SA_RESTART|SA_SIGINFO
    SIGTERM: UserHandler in libjvm.so, mask=11100100010111111101111111111110, flags=SA_RESTART|SA_SIGINFO
    SIGQUIT: UserHandler in libjvm.so, mask=11100100010111111101111111111110, flags=SA_RESTART|SA_SIGINFO
    SIGTRAP: crash_handler in libjvm.so, mask=11100100010111111101111111111110, flags=SA_RESTART|SA_SIGINFO
```

-----
Tests: GA, and the patch has been tested in our nighlies for over a month now. I manually executed the runtime/jni/checked tests too.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8260485](https://bugs.openjdk.java.net/browse/JDK-8260485): Simplify and unify handler vectors in Posix signal code


### Reviewers
 * [David Holmes](https://openjdk.java.net/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Gerard Ziemski](https://openjdk.java.net/census#gziemski) (@gerard-ziemski - Committer)


### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/2251/head:pull/2251`
`$ git checkout pull/2251`

dholmes-ora reviewed (changes requested):
Hi Thomas,

I've taken a first pass at this. I like parts of it, perhaps most of it, but am having trouble seeing the before/after picture clearly.

A few comments below.

Thanks,
David

tstuefe commented on src/hotspot/os/posix/signals_posix.cpp:
```diff
@@ -76,21 +87,75 @@ extern "C" {
+
+  bool check_signal_number(int sig) const {
+    assert(sig > 0 || sig < NSIG, "invalid signal number %d", sig);
+    return sig > 0 || sig < NSIG;
```
Good catch, this was an oversight.

tstuefe commented on src/hotspot/os/posix/signals_posix.cpp:
```diff
@@ -76,21 +87,75 @@ extern "C" {
+//  SavedSignalHandlers is a helper class for those cases, keeping an array of sigaction
+//  structures and membership information.
+class SavedSignalHandlers {
+  struct sigaction _v[NSIG];
```
v as in vector... I changed it to _sa.

tstuefe commented on src/hotspot/os/posix/signals_posix.cpp:
```diff
@@ -76,21 +87,75 @@ extern "C" {
+  bool is_set(int sig) const  { return sigismember(&_set, sig); }
+  void mark_as_set(int sig)   { sigaddset(&_set, sig); }
+  void mark_as_clear(int sig) { sigdelset(&_set, sig); }
+
```
True. I removed the whole sigset. See my comment below.

tstuefe commented on src/hotspot/os/posix/signals_posix.cpp:
```diff
@@ -777,64 +825,50 @@ static void check_signal_handler(int sig) {
+    if (sig == SHUTDOWN2_SIGNAL && !isatty(fileno(stdin))) {    // Flags?
       tty->print_cr("Running in non-interactive shell, %s handler is replaced by shell",
-                    os::exception_name(sig, buf, O_BUFLEN));
+                    os::exception_name(sig, buf, O_BUFLEN));    // When comparing, ignore the SA_RESTORER flag on Linux
```
Just a typo.

tstuefe commented:
> Hi Thomas,
> 
> I've taken a first pass at this. I like parts of it, perhaps most of it, but am having trouble seeing the before/after picture clearly.
> 
> A few comments below.
> 
> Thanks,
> David

Thanks Dave.

I reduced the complexity of the patch somewhat by removing case (3) completely. I found that it adds very little useful functionality. So all that is left now is (1) the simplified handling of chained handlers and (2) the new logic for implementing the signal handler jni check.

I also changed/simplified `SavedSignalHandlers` somewhat. NSIG can be largeish (I believe 1024 on AIX) and that would have made the array-of-sigaction structures large too. Not that it really matters much. But I changed the class into a pointer array, with the sigaction structures living in C-Heap. Since this array is sparsely populated, that saves space. Also removes the need for the membership sigset, since a NULL slot would indicate its not set.

Further answers inline. Feel free to ask questions if something is unclear.

Thanks, Thomas

dholmes-ora commented on src/hotspot/os/posix/signals_posix.cpp:
```diff
@@ -76,22 +87,74 @@ extern "C" {
+
+// For signal-chaining:
+//  if chaining is active, chained_handlers contains all handlers which we
+//  did replace with our own and to which we must delegate.
```
Nit: s/which we did replace/we replaced/

dholmes-ora commented on src/hotspot/os/posix/signals_posix.cpp:
```diff
@@ -777,64 +823,48 @@ static void check_signal_handler(int sig) {
     if (sig == SHUTDOWN2_SIGNAL && !isatty(fileno(stdin))) {
       tty->print_cr("Running in non-interactive shell, %s handler is replaced by shell",
-                    os::exception_name(sig, buf, O_BUFLEN));
+                    os::exception_name(sig, buf, O_BUFLEN));    // When comparing, ignore the SA_RESTORER flag on Linux
```
What does this comment mean in this context ???

tstuefe replied:
Nothing. This, like the "flags" comment before, must have been accidental paste errors. I went through the patch again, but I hope this was the last one.

dholmes-ora commented on src/hotspot/os/posix/signals_posix.cpp:
```diff
@@ -1346,53 +1370,51 @@ static void print_signal_set_short(outputStream* st, const sigset_t* set) {
-  }
+  st->print(", flags=");
+  int flags = act->sa_flags;
+  // On Linux, hide the SA_RESTORE flag
```
typo: SA_RESTORE -> SA_RESTORER

dholmes-ora commented on src/hotspot/os/posix/signals_posix.cpp:
```diff
@@ -1346,53 +1370,51 @@ static void print_signal_set_short(outputStream* st, const sigset_t* set) {
-  }
+// Print established signal handler for this signal.
+// - if this signal handler was installed by us and is chained to a pre-established user handler
+//    it did replace, print that one too.
```
s/it did replace/it replaced/

dholmes-ora commented on src/hotspot/os/posix/signals_posix.cpp:
```diff
@@ -1346,53 +1370,51 @@ static void print_signal_set_short(outputStream* st, const sigset_t* set) {
-      handler == CAST_FROM_FN_PTR(address, (sa_sigaction_t)SR_handler)) {
-    // It is our signal handler
-    // check for flags, reset system-used one!
-    if ((int)sa.sa_flags != get_our_sigflags(sig)) {
-      st->print(
-                ", flags was changed from " PTR32_FORMAT ", consider using jsig library",
-                get_our_sigflags(sig));
```
It is not clear to me that this check still remains somewhere.

tstuefe replied:
I removed this check since I did not think it very useful. We have the checking code with Xcheck:jni, which does the same check. 

But I am now reconsidering. If we run without Xcheck, it still is useful. I'll re-add this test.

dholmes-ora commented on src/hotspot/os/posix/vmError_posix.cpp:
```diff
@@ -49,16 +49,6 @@
-static const int NUM_SIGNALS = sizeof(SIGNALS) / sizeof(int);
-
-// Space for our "saved" signal flags and handlers
-static int resettedSigflags[NUM_SIGNALS];
-static address resettedSighandler[NUM_SIGNALS];
```
I'm not  clear what happened to these ?? (I'm not clear how they were used in the first place. :( )

tstuefe replied:
I think they were used for path (3). When printing signal handlers, we do this little test to check - similar to what we do in Xcheck:jni - to print out if someone changed the handler under us. See:
https://github.com/openjdk/jdk/blob/cb127a4bb5d95d19eb1f5e625b600311a2490135/src/hotspot/os/posix/signals_posix.cpp#L1372
and
https://github.com/openjdk/jdk/blob/cb127a4bb5d95d19eb1f5e625b600311a2490135/src/hotspot/os/posix/signals_posix.cpp#L1389
.
But when printing signal handlers as part of a hs-err file, we replaced the hotspot signal handlers with the secondary crash handler and would trigger this test. I think this whole logic is needed just to prevent this.

dholmes-ora reviewed (changes requested):
Hi Thomas,

Still a few minor comments and queries, but overall it is looking good.

Thanks,
David

tstuefe commented:
I added a printout to print_signal_handler to print out a message if the expected handler changed. This is the same logic, somewhat abridged, applied at Xcheck:jni. 

Only here, it fires whenever we print signal handlers independently from Xcheck:jni. So it gets also displayed in hs-err files or as part of a VM.info jcmd.

dholmes-ora commented on src/hotspot/os/posix/signals_posix.cpp:
```diff
@@ -118,6 +118,7 @@ class SavedSignalHandlers {
     if (check_signal_number(sig)) {
       assert(_sa[sig] == NULL, "Overwriting signal handler?");
       _sa[sig] = NEW_C_HEAP_OBJ(struct sigaction, mtInternal);
+      *_sa[sig] = *act;
```
Sorry I missed the fact we weren't actually copying the incoming sigaction struct!

tstuefe replied:
Yes me too, and I'm annoyed by the fact that no tier1 test did catch that. The full jtreg suite would have catched it but it did not run yet since the last commit.

dholmes-ora commented on src/hotspot/os/posix/signals_posix.cpp:
```diff
@@ -1409,6 +1410,25 @@ void PosixSignals::print_signal_handler(outputStream* st, int sig,
+      int this_flag = current_act.sa_flags;
+      LINUX_ONLY(this_flag &= SA_RESTORER_FLAG_MASK;)
+      if (this_flag != expected_act->sa_flags) {
+        st->print_cr("  *** Flags changed! Consider using jsig library.");
```
Can't we easily print the difference in flags?

tstuefe replied:
Sure. I'll change it.

dholmes-ora reviewed (approved):
Looks good. Thanks for the updates.

One query for further improvement below. :)

Thanks,
David

tstuefe commented:
> Looks good. Thanks for the updates.
> 
> One query for further improvement below. :)
> 
> Thanks,
> David

Thanks for approval, David. Unfortunately I still have a strange build problem on one of our s390 boxes, so this is not done yet.

tstuefe commented:
Hi David,

sorry, its not yet done.

Changes in this version:

1) I found that on zlinux, sigaction.sa_flags is actually a long unsigned int. Which surprised me since this is clearly not posix compatible. Therefore I rewrote the code, adding a new function `get_sanitized_sa_flags`, which hides the correct casting and also contains the stripping away of unwanted flags set by the kernel (SA_RESTORER). As a side effect this cleanly factors out the SA_RESTORER handling which is nice.

2) You requested that `print_single_signal_handler` should print the flag difference when spotting a change. I agree and expanded on this: now  `print_single_signal_handler` will print out the full expected handler information if it spots a difference (see below for examples). This contains more information that just the flags, and is actually less code.

3) Then I found that further code could be folded: We have a detailed printout for Xcheck:jni in `check_signal_handler`, but then we also print all signal handlers as part of that printout, which now (2) contains a verbose description of the found differences. Therefore the detailed printout in `check_signal_handler` was not needed anymore and could be removed. That also allows me to factor out the "compare sigaction structures semantically" logic into one function, called `compare_handler_info`. That again safes some code.

Now the logic is like this:

- If Xcheck:jni is not set, we still store the expected signal handler info. When printing signal handlers, e.g. as part of a hs_err file, in `os::print_signal_handlers`, we print a description of any changed handlers to nudge the supporter into the right direction.

- If Xcheck:jni is active, we will do our periodic checks; upon finding a difference, we do not elaborate but call `os::print_signal_handlers()` which will elaborate for us.

The final adjustment I had to make was to separate the "expected handler" info from the "check when Xcheck:jni is active" logic. That is because when Xcheck:jni is active and we find a mismatch, we want to disable checking for that signal in `check_signal_handler` but still see the difference in future printouts when calling `os::print_signal_handlers`, e.g. if later we crash and write a hs-err file.

I am really sorry for the added review work, but I think its a worthwhile change and complexity gets further reduced.

I thought about adding tests but testing this is quite difficult. So I tested manually by overwriting some of the hotspot signal handlers and check how the VM reacts.

First a hs-err file, good case:

```
659 Signal Handlers:
660    SIGSEGV: crash_handler in libjvm.so, mask=11100100010111111101111111111110, flags=SA_RESTART|SA_SIGINFO
661     SIGBUS: crash_handler in libjvm.so, mask=11100100010111111101111111111110, flags=SA_RESTART|SA_SIGINFO
662     SIGFPE: crash_handler in libjvm.so, mask=11100100010111111101111111111110, flags=SA_RESTART|SA_SIGINFO
663    SIGPIPE: javaSignalHandler in libjvm.so, mask=11100100010111111101111111111110, flags=SA_RESTART|SA_SIGINFO
664    SIGXFSZ: javaSignalHandler in libjvm.so, mask=11100100010111111101111111111110, flags=SA_RESTART|SA_SIGINFO
665     SIGILL: crash_handler in libjvm.so, mask=11100100010111111101111111111110, flags=SA_RESTART|SA_SIGINFO
666    SIGUSR2: SR_handler in libjvm.so, mask=00000000000000000000000000000000, flags=SA_RESTART|SA_SIGINFO
667     SIGHUP: UserHandler in libjvm.so, mask=11100100010111111101111111111110, flags=SA_RESTART|SA_SIGINFO
668     SIGINT: UserHandler in libjvm.so, mask=11100100010111111101111111111110, flags=SA_RESTART|SA_SIGINFO
669    SIGTERM: UserHandler in libjvm.so, mask=11100100010111111101111111111110, flags=SA_RESTART|SA_SIGINFO
670    SIGQUIT: UserHandler in libjvm.so, mask=11100100010111111101111111111110, flags=SA_RESTART|SA_SIGINFO
671    SIGTRAP: crash_handler in libjvm.so, mask=11100100010111111101111111111110, flags=SA_RESTART|SA_SIGINFO
```

Bad cases (I changed handlers for SIGFPE, SIGPIPE, SIGXFSZ and SIGUSR2):

hs-err file:

```
658 Signal Handlers:
659    SIGSEGV: crash_handler in libjvm.so, mask=11100100010111111101111111111110, flags=SA_RESTART|SA_SIGINFO
660     SIGBUS: crash_handler in libjvm.so, mask=11100100010111111101111111111110, flags=SA_RESTART|SA_SIGINFO
661     SIGFPE: badwolf in libjvm.so, mask=10000000000000000000000000000000, flags=none
662   *** Handler was modified!
663   *** Expected: javaSignalHandler in libjvm.so, mask=11100100110111111111111111111110, flags=SA_RESTART|SA_SIGINFO
664    SIGPIPE: badwolf in libjvm.so, mask=10000000000000000000000000000000, flags=none
665   *** Handler was modified!
666   *** Expected: javaSignalHandler in libjvm.so, mask=11100100110111111111111111111110, flags=SA_RESTART|SA_SIGINFO
667    SIGXFSZ: badwolf in libjvm.so, mask=10000000000000000000000000000000, flags=none
668   *** Handler was modified!
669   *** Expected: javaSignalHandler in libjvm.so, mask=11100100110111111111111111111110, flags=SA_RESTART|SA_SIGINFO
670     SIGILL: crash_handler in libjvm.so, mask=11100100010111111101111111111110, flags=SA_RESTART|SA_SIGINFO
671    SIGUSR2: badwolf in libjvm.so, mask=10000000000000000000000000000000, flags=none
672   *** Handler was modified!
673   *** Expected: SR_handler in libjvm.so, mask=00000000000000000000000000000000, flags=SA_RESTART|SA_SIGINFO
674     SIGHUP: UserHandler in libjvm.so, mask=11100100010111111101111111111110, flags=SA_RESTART|SA_SIGINFO
675     SIGINT: UserHandler in libjvm.so, mask=11100100010111111101111111111110, flags=SA_RESTART|SA_SIGINFO
676    SIGTERM: UserHandler in libjvm.so, mask=11100100010111111101111111111110, flags=SA_RESTART|SA_SIGINFO
677    SIGQUIT: UserHandler in libjvm.so, mask=11100100010111111101111111111110, flags=SA_RESTART|SA_SIGINFO
678    SIGTRAP: crash_handler in libjvm.so, mask=11100100010111111101111111111110, flags=SA_RESTART|SA_SIGINFO
```

jcmd VM.info:

```
...
Signal Handlers:
   SIGSEGV: javaSignalHandler in libjvm.so, mask=11100100010111111101111111111110, flags=SA_RESTART|SA_SIGINFO
    SIGBUS: javaSignalHandler in libjvm.so, mask=11100100010111111101111111111110, flags=SA_RESTART|SA_SIGINFO
    SIGFPE: badwolf in libjvm.so, mask=00000111001010000101001100010001, flags=none
  *** Handler was modified!
  *** Expected: javaSignalHandler in libjvm.so, mask=11100100110111111111111111111110, flags=SA_RESTART|SA_SIGINFO
   SIGPIPE: badwolf in libjvm.so, mask=00000111001010000101001100010001, flags=none
  *** Handler was modified!
  *** Expected: javaSignalHandler in libjvm.so, mask=11100100110111111111111111111110, flags=SA_RESTART|SA_SIGINFO
   SIGXFSZ: badwolf in libjvm.so, mask=00000111001010000101001100010001, flags=none
  *** Handler was modified!
  *** Expected: javaSignalHandler in libjvm.so, mask=11100100110111111111111111111110, flags=SA_RESTART|SA_SIGINFO
    SIGILL: javaSignalHandler in libjvm.so, mask=11100100010111111101111111111110, flags=SA_RESTART|SA_SIGINFO
   SIGUSR2: badwolf in libjvm.so, mask=00000111001010000101001100010001, flags=none
  *** Handler was modified!
  *** Expected: SR_handler in libjvm.so, mask=00000000000000000000000000000000, flags=SA_RESTART|SA_SIGINFO
    SIGHUP: UserHandler in libjvm.so, mask=11100100010111111101111111111110, flags=SA_RESTART|SA_SIGINFO
    SIGINT: UserHandler in libjvm.so, mask=11100100010111111101111111111110, flags=SA_RESTART|SA_SIGINFO
   SIGTERM: UserHandler in libjvm.so, mask=11100100010111111101111111111110, flags=SA_RESTART|SA_SIGINFO
   SIGQUIT: UserHandler in libjvm.so, mask=11100100010111111101111111111110, flags=SA_RESTART|SA_SIGINFO
   SIGTRAP: SIG_DFL, mask=00000000000000000000000000000000, flags=none
```


-Xcheck:jni output:

```
Warning: SIGFPE handler modified!                                                                                                                                                                                                                                                    
Signal Handlers:
   SIGSEGV: javaSignalHandler in libjvm.so, mask=11100100010111111101111111111110, flags=SA_RESTART|SA_SIGINFO
    SIGBUS: javaSignalHandler in libjvm.so, mask=11100100010111111101111111111110, flags=SA_RESTART|SA_SIGINFO
    SIGFPE: badwolf in libjvm.so, mask=11100100010111111101111111111110, flags=none
  *** Handler was modified!
  *** Expected: javaSignalHandler in libjvm.so, mask=11100100110111111111111111111110, flags=SA_RESTART|SA_SIGINFO
   SIGPIPE: badwolf in libjvm.so, mask=11100100010111111101111111111110, flags=none
  *** Handler was modified!
  *** Expected: javaSignalHandler in libjvm.so, mask=11100100110111111111111111111110, flags=SA_RESTART|SA_SIGINFO
   SIGXFSZ: badwolf in libjvm.so, mask=11100100010111111101111111111110, flags=none
  *** Handler was modified!
  *** Expected: javaSignalHandler in libjvm.so, mask=11100100110111111111111111111110, flags=SA_RESTART|SA_SIGINFO
    SIGILL: javaSignalHandler in libjvm.so, mask=11100100010111111101111111111110, flags=SA_RESTART|SA_SIGINFO
   SIGUSR2: badwolf in libjvm.so, mask=11100100010111111101111111111110, flags=none
  *** Handler was modified!
  *** Expected: SR_handler in libjvm.so, mask=00000000000000000000000000000000, flags=SA_RESTART|SA_SIGINFO
    SIGHUP: UserHandler in libjvm.so, mask=11100100010111111101111111111110, flags=SA_RESTART|SA_SIGINFO
    SIGINT: UserHandler in libjvm.so, mask=11100100010111111101111111111110, flags=SA_RESTART|SA_SIGINFO
   SIGTERM: UserHandler in libjvm.so, mask=11100100010111111101111111111110, flags=SA_RESTART|SA_SIGINFO
   SIGQUIT: UserHandler in libjvm.so, mask=11100100010111111101111111111110, flags=SA_RESTART|SA_SIGINFO
   SIGTRAP: SIG_DFL, mask=00000000000000000000000000000000, flags=none
Consider using jsig library.
...
```
Thanks, Thomas

dholmes-ora commented on src/hotspot/os/posix/signals_posix.cpp:
```diff
@@ -141,12 +146,12 @@ class SavedSignalHandlers {
+//  and compare it periodically against reality (see os::run_periodic_checks()).
 static bool check_signals = true;
 static SavedSignalHandlers expected_handlers;
+static bool do_check_signal_periodically[NSIG];
```
Does this need explicit initialization to have false entries?

tstuefe replied:
I think it does not since global static objects are zero-initialized and 0 = false; however, I made the initialization explicit.

dholmes-ora reviewed (approved):
Still seems okay.

One query below.

Thanks,
David

tstuefe commented:
> Still seems okay.
> 
> One query below.
> 
> Thanks,
> David

Thank you David.

tstuefe commented:
Gentle ping..

gerard-ziemski commented:
hi Thomas, I'm interested in reviewing your fix and will work on it as soon as I'm done with https://github.com/openjdk/jdk/pull/2403

(tomorrow?)

gerard-ziemski commented:
I'm starting the review...

gerard-ziemski commented on src/hotspot/os/posix/signals_posix.cpp:
```diff
@@ -76,22 +99,68 @@ extern "C" {
+  // this is no problem.
+  struct sigaction* _sa[NSIG];
+
+  bool check_signal_number(int sig) const {
+    assert(sig > 0 && sig < NSIG, "invalid signal number %d", sig);
+    return sig > 0 && sig < NSIG;
+  }
```
We have

```c
 assert(sig > 0 && sig < NSIG, "invalid signal number %d", sig);
 return sig > 0 && sig < NSIG;
```

here in `SavedSignalHandlers::check_signal_number()` and

```c
#if defined(__APPLE__)
  return sig >= 1 && sig < NSIG;
#else
```

in `is_valid_signal()`.

Can we combine those, something like:

```c
// Returns true if signal number is valid.
static bool is_valid_signal(int sig) {
  assert(sig > 0 && sig < NSIG, "invalid signal number %d", sig);
  if (sig > 0 && sig < NSIG) {
    // Use sigaddset to check for signal validity.
    sigset_t set;
    sigemptyset(&set);
    if (sigaddset(&set, sig) == -1 && errno == EINVAL) {
      return false;
    } else {
      return true;
    }
  } else {
    return false;
  }
}
```

so then we can drop `SavedSignalHandlers::check_signal_number()` altogether?

tstuefe replied:
I would like to keep this. SavedSignalHandlers::check_signal_number() is just a simple bounds check to prevent memory overwriters, with an added assert for goot measure in debug builds. Its proximity to the array it guards makes this clear. SavedSignalHandlers gets only fed with know values in this file, so an assert is sufficient.

is_valid_signal() OTOH does a runtime check for signal validity. It is more expensive but only used in analysis situations. It pulls its signal number from a sigaction structure, which may be corrupt and contain whatever, so the runtime check makes more sense to me.

tstuefe replied:
In fact, SavedSignalHandler::_sa[NSIG] could be reduced in size to something more sensible, since we only ever feed "normal" signals into it. Same reason as NUM_IMPORTANT_SIGS. But we only talk a few kbyte here, so I did not bother.

gerard-ziemski replied:
I would have preferred to have a single function to check whether the signal is valid, otherwise we have 2 separate specialized API that almost do the same thing.  Any future reader of the code will probably wonder again.

For code readability I'd prefer to have just one API, or at least refactor the more complex one in terms of the simpler one, and give them meaningful names to differentiate them, but OK.

gerard-ziemski commented on src/hotspot/os/posix/signals_posix.cpp:
```diff
@@ -357,7 +411,7 @@ struct sigaction* get_chained_signal_action(int sig) {
   if (actp == NULL) {
     // Retrieve the preinstalled signal handler from jvm
-    actp = get_preinstalled_handler(sig);
+    actp = const_cast<struct sigaction*>(chained_handlers.get(sig));
```
Must `SavedSignalHandlers::get()` have the **const struct** in `const struct sigaction* get(int sig) const` signature?

If it was just `struct sigaction* get(int sig) const` then we wouldn't need this awkward cast.

tstuefe replied:
I'm a stickler to const since its very expressive and prevents stupid errors. `const struct sigaction* get() const` makes perfect sense since at no time you are supposed to modify the underlying sigaction structure. Once it is saved, it should only be used to read from (eg to print, or to extract the handler address and call it in `call_chained_handler()`.

But old code was not const correct. The way to start being const correct without having to rewrite everything is to make new code const correct and cauterize at the interface between new and old code with these ugly but at least expressive casts.

Side note, even the sigaction function is const correct: https://pubs.opengroup.org/onlinepubs/007904875/functions/sigaction.html
see the `act` input parameter designating the new handler, its const. Only the output parameter `oact` is nonconst.

I will see if I can make the rest of the code const correct too wrt the sigaction structures. That way we can get rid of the cast and have better code as well.

gerard-ziemski replied:
If we can do that, then that would be ideal.

gerard-ziemski reviewed (changes requested):
I did not get as far as I hoped today, I'll review more tomorrow.

gerard-ziemski commented on src/hotspot/os/posix/signals_posix.cpp:
```diff
@@ -1203,16 +1217,17 @@ void set_signal_handler(int sig) {
 
+  // Query the current signal handler. Needs to be a separate operation
+  // from installing a new handler since we need to honor AllowUserSignalHandlers.
   void* oldhand = get_signal_handler(&oldAct);
```
It's a pre-existing issue, but I dislike the **"old"** in **oldhand** and **oldAct**. It makes it sound like it is a cached or previous value, which is especially confusing when in the comment we refer to it as the **current** one. Any chance we can clean this up in this fix?

tstuefe replied:
I think it stems from the Posix interface calling the parameters `act` and `oact` and ppl are used to that naming:
```
int sigaction(int sig, const struct sigaction *restrict act,
       struct sigaction *restrict oact);
```
That said, we use a bunch of different names, like oact, oldAct or oldSigAct, which I disliked. I wanted to unify those namings but refrained since I wanted to keep the patch small. Its difficult enough as it is.

Since this would be an easy change in itself, can we keep it for a different RFE? As for name, lets say currentAct?

gerard-ziemski replied:
I'd love if it we could follow up on it.

gerard-ziemski commented on src/hotspot/os/posix/signals_posix.cpp:
```diff
@@ -76,22 +99,68 @@ extern "C" {
+//  To check that, and to aid with diagnostics, store a copy of the handler setup
+//  and compare it periodically against reality (see os::run_periodic_checks()).
+static bool check_signals = true;
+static SavedSignalHandlers expected_handlers;
```
I personally would prefer `vm_handlers` or `java_handlers` or `our_handlers` here instead of `expected_handlers`.

tstuefe replied:
Okay, vm_handlers it is.

gerard-ziemski commented on src/hotspot/os/posix/signals_posix.cpp:
```diff
@@ -763,10 +805,32 @@ typedef int (*os_sigaction_t)(int, const struct sigaction *, struct sigaction *)
+  address expected_handler = get_signal_handler(expected_sa);
+  const int this_flags = get_sanitized_sa_flags(sa);
+  const int expected_flags = get_sanitized_sa_flags(expected_sa);
+  return this_handler == expected_handler &&
+         this_flags == expected_flags;
```
Could we use brackets here like:

```c
  return ((this_handler == expected_handler) &&
           (this_flags == expected_flags));
```

tstuefe replied:
I can add the inside brackets, but don't see the need for the outside ones, there is nothing to bracket off against.

gerard-ziemski commented on src/hotspot/os/posix/signals_posix.cpp:
```diff
@@ -777,65 +841,22 @@ static void check_signal_handler(int sig) {
-    // No need to check this sig any longer
-    sigaddset(&check_signal_done, sig);
+  // Compare both sigaction structures (intelligently; only the members we care about).
+  if (!compare_handler_info(&act, expected_act)) {
```
The name `compare_handler_info()` makes it sound like a function that tells whether we should compare the values, not whether the values' comparison evaluates to true.

How about `are_handlers_equal()`?

tstuefe replied:
Ok

gerard-ziemski commented on src/hotspot/os/posix/signals_posix.cpp:
```diff
@@ -777,65 +841,22 @@ static void check_signal_handler(int sig) {
+  if (!compare_handler_info(&act, expected_act)) {
+    tty->print_cr("Warning: %s handler modified!", os::exception_name(sig, buf, sizeof(buf)));
+    // If we had a mismatch:
+    // - print all signal handlers. As part of that printout details will be printed
```
Needs comma:

`// - print all signal handlers. As part of that printout, details will be printed`

gerard-ziemski commented on src/hotspot/os/posix/signals_posix.cpp:
```diff
@@ -1239,15 +1254,16 @@ void set_signal_handler(int sig) {
 
   void* oldhand2  = get_signal_handler(&oldAct);
   assert(oldhand2 == oldhand, "no concurrent signal handler installation");
+
```
Empty line.

gerard-ziemski commented:
General comment: I get weird results when I try to run `java -XX:ErrorHandlerTest=2 -version`

With the proposed changes, I get:

```c
oracle@dhcp-10-154-126-49 jdk % ./build/xcode/build/jdk/bin/java -XX:ErrorHandlerTest=2 -version
# To suppress the following error report, specify this argument
# after -XX: or in .hotspotrc:  SuppressErrorAt=/vmError.cpp:1785
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  


[Too many errors, abort]
zsh: abort      ./build/xcode/build/jdk/bin/java -XX:ErrorHandlerTest=2 -version
```

with several empty lines, which I deleted for compactness, whereas the original code produces:

```c
oracle@Oracles-MacBook-Pro-16 jdk_orig % ./build/xcode/build/jdk/bin/java -XX:ErrorHandlerTest=2 -version
# To suppress the following error report, specify this argument
# after -XX: or in .hotspotrc:  SuppressErrorAt=/vmError.cpp:1785
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  Internal Error (/Volumes/Work/review/2251/jdk_orig/src/hotspot/share/utilities/vmError.cpp:1785), pid=69509, tid=7427
#  guarantee(how == 0) failed: test guarantee
#
# JRE version: OpenJDK Runtime Environment (17.0) (fastdebug build 17-internal+0-adhoc.oracle.jdkorig)
# Java VM: OpenJDK 64-Bit Server VM (fastdebug 17-internal+0-adhoc.oracle.jdkorig, mixed mode, tiered, compressed oops, compressed class ptrs, g1 gc, bsd-amd64)
# No core dump will be written. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
#
# An error report file with more information is saved as:
# /Volumes/Work/review/2251/jdk_orig/hs_err_pid69509.log
#
# If you would like to submit a bug report, please visit:
#   https://bugreport.java.com/bugreport/crash.jsp
#
zsh: abort      ./build/xcode/build/jdk/bin/java -XX:ErrorHandlerTest=2 -version
```

so something is off.

tstuefe commented:
> java -XX:ErrorHandlerTest=2 -version

I cannot reproduce it. What build is this? The error is weird since "too many errors, abort" should have been preceded by any number of secondary error printouts.

gerard-ziemski commented:
> > java -XX:ErrorHandlerTest=2 -version
> 
> I cannot reproduce it (macos, fastdebug). What build is this? The error is weird since "too many errors, abort" should have been preceded by any number of secondary error printouts.
> 
> These kind of tests (let it crash and check if error handling works) are also executed as part of tier1 hotspot (runtime/ErrorHandling) and seem to run fine in GA actions.

I'll look into why I see this and report back...

tstuefe commented:
> > > java -XX:ErrorHandlerTest=2 -version
> > 
> > 
> > I cannot reproduce it (macos, fastdebug). What build is this? The error is weird since "too many errors, abort" should have been preceded by any number of secondary error printouts.
> > These kind of tests (let it crash and check if error handling works) are also executed as part of tier1 hotspot (runtime/ErrorHandling) and seem to run fine in GA actions.
> 
> I'll look into why I see this and report back...

Okay, thank you. Note that I'm off the rest of the week, so I won't be able to reply until Monday.

gerard-ziemski commented:
It looks like I see the issue only with Xcode build hotspot (using the actual Xcode IDE), the build produced using the command line make/compiler works fine. Digging in deeper...

dholmes-ora reviewed (approved):
Still good to me.

Thanks,
David

tstuefe commented:
> Still good to me.
> 
> Thanks,
> David

Thanks, David

gerard-ziemski commented:
I'm going to give thumbs up and not hold this up any longer, even though I haven't figured out yet why I see the issue with my Xcode built hotspot.

Whatever the issue, which is most likely due to how I build it, it can be handled either on my side, or if it's JDK then we can fix it in a followup.

cheers

tstuefe commented:
Thanks Gerard and David!

About const correctness, I looked into this. Looked mostly straightforward, apart from the one fact that inside call_chained_handler, we modify the handed over sigaction structure. Which is really yucky. Especially if you consider that this function is called also for the libjsig case, where the sigaction structure is handed in from the libjsig, so we essentially modify memory in libjsig.

We do this for two reasons, one is to just have a holder for a temporary sigset, which would be trivial to fix with a temporary local variable. And then, to switch off chaining for SA_NODEFER (mimicking one shot semantics). And while looking into that, I saw that one shot semantics can never have worked reliably. I guess its like David said in other places, its a best effort thing.

The whole investigation turned out too deep for this RFE, so I leave that for another day (https://bugs.openjdk.java.net/browse/JDK-8262006).

@gerard-ziemski : I also created https://bugs.openjdk.java.net/browse/JDK-8262007 to track the renaming you wanted. Left it unassigned, feel free to grab that one.

Cheers, Thomas

/integrate
